File: Nerava/Nerava/Engine/SessionState.swift
import Foundation

// MARK: - Session State

enum SessionState: String, Codable {
    case idle = "IDLE"
    case nearCharger = "NEAR_CHARGER"
    case anchored = "ANCHORED"
    case sessionActive = "SESSION_ACTIVE"
    case inTransit = "IN_TRANSIT"
    case atMerchant = "AT_MERCHANT"
    case sessionEnded = "SESSION_ENDED"
}

// MARK: - Session Events (Canonical - No Duplicates)

/// Each semantic event has exactly ONE enum case.
/// Use `requiresSessionId` to determine routing (pre-session vs session endpoint).
enum SessionEvent: String, Codable {
    // Pre-session events (no session_id yet)
    case chargerTargeted = "charger_targeted"
    case enteredChargerIntentZone = "entered_charger_intent_zone"
    case exitedChargerIntentZone = "exited_charger_intent_zone"
    case anchorDwellComplete = "anchor_dwell_complete"
    case anchorLost = "anchor_lost"
    case activationRejected = "activation_rejected"

    // Session events (has session_id)
    case exclusiveActivatedByWeb = "exclusive_activated"
    case departedCharger = "departed_charger"
    case enteredMerchantZone = "entered_merchant_zone"
    case visitVerifiedByWeb = "visit_verified"
    case gracePeriodExpired = "grace_period_expired"
    case hardTimeoutExpired = "hard_timeout_expired"
    case webRequestedEnd = "web_requested_end"
    case sessionRestored = "session_restored"

    /// Determines whether this event requires a session_id (routes to session endpoint)
    /// or can be sent without one (routes to pre-session endpoint).
    var requiresSessionId: Bool {
        switch self {
        case .chargerTargeted,
             .enteredChargerIntentZone,
             .exitedChargerIntentZone,
             .anchorDwellComplete,
             .anchorLost,
             .activationRejected:
            return false
        case .exclusiveActivatedByWeb,
             .departedCharger,
             .enteredMerchantZone,
             .visitVerifiedByWeb,
             .gracePeriodExpired,
             .hardTimeoutExpired,
             .webRequestedEnd,
             .sessionRestored:
            return true
        }
    }
}

// MARK: - Supporting Types

struct ChargerTarget: Codable, Equatable {
    let id: String
    let latitude: Double
    let longitude: Double
}

struct MerchantTarget: Codable, Equatable {
    let id: String
    let latitude: Double
    let longitude: Double
}

struct ActiveSessionInfo: Codable, Equatable {
    let sessionId: String
    let chargerId: String
    let merchantId: String
    let startedAt: Date
}

/// Full pending event payload for retry on restore.
/// Persisted in snapshot so we can resend if crash/kill before ack.
struct PendingEvent: Codable, Equatable {
    let eventId: String
    let eventName: String
    let requiresSessionId: Bool
    let sessionId: String?
    let chargerId: String?
    let occurredAt: Date
    let metadata: [String: String]  // Simplified for Codable
}

File: mobile/nerava_android/app/src/main/java/network/nerava/app/engine/SessionState.kt
package network.nerava.app.engine

/**
 * Mirror of iOS SessionState enum.
 * Raw values match the strings sent over the bridge and to the backend.
 */
enum class SessionState(val raw: String) {
    IDLE("IDLE"),
    NEAR_CHARGER("NEAR_CHARGER"),
    ANCHORED("ANCHORED"),
    SESSION_ACTIVE("SESSION_ACTIVE"),
    IN_TRANSIT("IN_TRANSIT"),
    AT_MERCHANT("AT_MERCHANT"),
    SESSION_ENDED("SESSION_ENDED");

    companion object {
        fun fromRaw(raw: String): SessionState? = entries.find { it.raw == raw }
    }
}

/**
 * Mirror of iOS SessionEvent enum.
 * Each event has exactly one enum case. Use requiresSessionId to route to the correct endpoint.
 */
enum class SessionEvent(val raw: String, val requiresSessionId: Boolean) {
    // Pre-session events (no session_id)
    CHARGER_TARGETED("charger_targeted", false),
    ENTERED_CHARGER_INTENT_ZONE("entered_charger_intent_zone", false),
    EXITED_CHARGER_INTENT_ZONE("exited_charger_intent_zone", false),
    ANCHOR_DWELL_COMPLETE("anchor_dwell_complete", false),
    ANCHOR_LOST("anchor_lost", false),
    ACTIVATION_REJECTED("activation_rejected", false),

    // Session events (has session_id)
    EXCLUSIVE_ACTIVATED("exclusive_activated", true),
    DEPARTED_CHARGER("departed_charger", true),
    ENTERED_MERCHANT_ZONE("entered_merchant_zone", true),
    VISIT_VERIFIED("visit_verified", true),
    GRACE_PERIOD_EXPIRED("grace_period_expired", true),
    HARD_TIMEOUT_EXPIRED("hard_timeout_expired", true),
    WEB_REQUESTED_END("web_requested_end", true),
    SESSION_RESTORED("session_restored", true),
}

data class ChargerTarget(
    val id: String,
    val latitude: Double,
    val longitude: Double,
)

data class MerchantTarget(
    val id: String,
    val latitude: Double,
    val longitude: Double,
)

data class ActiveSessionInfo(
    val sessionId: String,
    val chargerId: String,
    val merchantId: String,
    val startedAt: Long, // epoch millis
)
